Draft: CIP Delegate-less Automation

julientinguely-da opened the pull request.

julientinguely-da committed:
init

julientinguely-da committed:
typos`

julientinguely-da committed:
typos

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+## Abstract
+
+This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 
+Instead of relying on a single delegate, we propose an approach where all SVs attempt to exercise delegate choices. 
```
```suggestion
Instead of relying on a single delegate, we propose an approach where all SV application attempt to exercise delegate choices. 
```

(to be clear that this is not manual)

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+
+This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 
+Instead of relying on a single delegate, we propose an approach where all SVs attempt to exercise delegate choices. 
+Each applying a random delay tuned to minimize contention.
```
```suggestion
Each application will apply a random delay, tuned to minimize contention.
```

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+
+### Remove Delegate Election
+
+In the current implementation, there is a dedicated page to elect a new delegate for the next round, 
```
```suggestion
In the current implementation, there is a dedicated page in the UI to elect a new delegate for the next round, 
```

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+
+## Motivation
+
+A delegate performs critical network-related tasks, such as for example advancing mining rounds or close vote proposals. 
```
```suggestion
A delegate performs critical network-related tasks, such as for example advancing mining rounds or acting on approved vote proposals. 
```

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+In the current version, many other core processes rely heavily on the delegate, 
+with no built-in fault tolerance to mitigate issues that may occur on the node. 
+A concerning problem observed multiple times is when a malfunctioning delegate causes mining rounds to stop progressing altogether. 
+This has raised concerns among SV operators and highlights the need for an better fault tolerance mechanism.
```
```suggestion
This has raised concerns among SV operators and highlights the need for a better fault tolerance mechanism.
```

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+  all automation should continue to function without significant impact on latency or throughput of tasks processed
+
+* **Minimize Contention:** While some contention are likely unavoidable, it should be minimized. 
+  Given that SVs do not pay traffic contention is not a problem wrt to traffic costs, but it does put additional load on the network.
```
```suggestion
  Given that SVs do not pay for traffic, contention is not a problem wrt to traffic costs, but it does put additional load on the network.
```

isegall-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,87 @@
+
+## Rationale
+
+### Requirements
```
I believe the requirements belong in motivation, and then the rationale should explain how the Specification satisfies the requirements.

julientinguely-da replied:
Thanks for the review Itai! I was not sure how to integrate these requirements. Let me iterate again on that.

moritzkiefer-da replied:
I'd honestly just drop the requirements section

julientinguely-da replied:
dropping

julientinguely-da committed:
review Itai 1

julientinguely-da committed:
review Itai 2: move requirements to motivation, rationale explains how specification solve requirements

julientinguely-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+To avoid SVs attempting a task at the same time and having only one succeeding, 
+we are adding a random delay before each attempt and tweak the canton retry backoff parameter.
+Each random delay D is drawn from a uniform distribution D ~ Uniform(0, U), where the upper bound U is max(n * t, T) 
+with n = number of SVs, t = estimated task duration, T = polling trigger interval (30 seconds).
```
missing: should talk about how we tune retries _initialDelay_ and _maxDelay_ so that we don't run into too much contention. Probably increasing the _intialDelay_ is enough here.

julientinguely-da replied:
Correction: U is max(n * t, T - t)

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+
+### All SVs try to exercise delegate choices
+
+In the new implementation, all current choices in _dsoRules_ having _dsoDelegate_ as controller will introduce 
```
```suggestion
In the new implementation, all current choices on the `DsoRules` template having `dsoDelegate` as controller will introduce 
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+### All SVs try to exercise delegate choices
+
+In the new implementation, all current choices in _dsoRules_ having _dsoDelegate_ as controller will introduce 
+a new optional argument named _sv_ to be used instead.
```
```suggestion
a new optional argument named `sv` to be used instead. This allows all SVs to exercise the choice by specifying their own party. Daml will reject a transaction where they try to specify any party other than their own with an authorization error.
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+
+In the new implementation, all current choices in _dsoRules_ having _dsoDelegate_ as controller will introduce 
+a new optional argument named _sv_ to be used instead.
+When the new _delegatelessAutomation_ feature flag is set and the _splice-dso-governance.dar_ implementing this change is vetted, 
```
```suggestion
Once this CIP is implemented,
```

The feature flag is for us during development, we shouldn't mention that externally.

julientinguely-da replied:
makes sense

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+a new optional argument named _sv_ to be used instead.
+When the new _delegatelessAutomation_ feature flag is set and the _splice-dso-governance.dar_ implementing this change is vetted, 
+SV applications will try to execute these choices by passing their own SV PartyId as controller. 
+The main challenge is to avoid contention - concurrent executions of the same choice.
```
```suggestion
```
the next section describes this already

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+Each random delay D is drawn from a uniform distribution D ~ Uniform(0, U), where the upper bound U is max(n * t, T) 
+with n = number of SVs, t = estimated task duration, T = polling trigger interval (30 seconds).
+
+A new metric named `splice_trigger_attempted_total` is added to measure and monitor trigger contention.
```
```suggestion
### SV App

The sv app is modified such that each SV attempts to submit the commands currently only submitted by the delegate.

This introduces some potential for conflict: If multiple SVs submit the same transaction at the same time, e.g., all attempt to advance the mining rounds at the same time, only one of them will succeed and the others fail, i.e., they contend with each other. Importantly, this is not a correctness issue: The Daml code already has to handle this case, e.g., if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.

To minimize the impact, each SV adds a random delay before each attempt and retries with a random backoff.

A new metric named `splice_trigger_attempted_total` is added to measure and monitor trigger contention and chose the right delay and backoff parameters.
```

moritzkiefer-da replied:
I think the details of the backoff are not all that interesting for the CIP. The important part is:
1. None of this matters for correctness, it's only about minimizing load and latency.
2. It's monitored and can be tweaked.

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+In the current implementation, there is a dedicated page in the SV UI to elect a new delegate for the next round, 
+intended to unlock the system if a delegate comes down.
+In the new implementation, delegate election will be fully removed, 
+associated Daml choices marked as deprecated and the _Delegate Election_ page in the SV UI removed.
```
```suggestion
associated Daml choices cannot be executed anymore and the _Delegate Election_ page in the SV UI removed.
```

imho the deprecated part is what we do for now during development. Eventually we should just nuke them (i.e. change the implementation to abort)

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+
+## Motivation
+
+A delegate performs critical network-related tasks, such as for example advancing mining rounds or acting on voted proposals. 
```
```suggestion
A delegate performs critical network-related tasks, such as for example advancing mining rounds or executing accepted votes.
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+
+A delegate performs critical network-related tasks, such as for example advancing mining rounds or acting on voted proposals. 
+In the current version, many other core processes rely heavily on the delegate, 
+with no built-in fault tolerance to mitigate issues that may occur on the node. 
```
```suggestion
In the current version, there is no automatic fault tolerance to mitigate issues that may occur on the node. SVs can vote on a new delegate but this is a manual process and just changes the single point of failure to a different SV.
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+In the current version, many other core processes rely heavily on the delegate, 
+with no built-in fault tolerance to mitigate issues that may occur on the node. 
+A concerning problem observed multiple times is when a malfunctioning delegate causes mining rounds to stop progressing altogether. 
+This has raised concerns among SV operators and highlights the need for a better fault tolerance mechanism.
```
```suggestion
This has been observed several times on devnet and testnet and highlights the need for a better fault tolerance mechanism.
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,93 @@
+
+## Rationale
+
+### Our new approach
+
+Having all SV applications attempting tasks adds tolerance against faulty and malicious SVs as long as a 2/3 super majority is trustful.
+Adding random delays and tweaking backoff retries minimizes contention.
+The new metric helps with accountability and tuning our parameters.
+
+### Randomization vs. Task-based seeds
```
I think this section leaks the design doc too much. We don't talk about task based seeds at all here so this title makes no sense.

I would just mention something like

"All SVs attempting to complete tasks provides the right fault tolerance guarantees and introduces no correctness issues as commands are already all idempotent. Random delays and backoffs are sufficient to minimize contention given that the delegate based choices don't require high throughput, e.g., round advancement happens only every 10min.

While more complex approaches to minimize contention would be possible, e.g., defining a clear order in which SVs should attempt to process a given task that is determined based on some task identifier, the implementation complexity of that does not seem to be justified.

julientinguely-da replied:
ok makes sense again, replacing the current Rationale section with your suggestion.

julientinguely-da committed:
review moritz

julientinguely-da committed:
self review

julientinguely-da commented:
Thanks for the great reviews! I took your suggestions and did some tiny modifications.
If you're good with it @isegall-da and @moritzkiefer-da, I'll open a PR with this CIP against main (+the push the branch on splice).

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+
+### SV Application
+
+The SV application is modified such that it attempts to submit the commands currently only being submitted by the delegate.
```
```suggestion
The SV application is modified such that all SVs attempt to submit the commands currently only being submitted by the delegate.
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+In the new implementation, all current choices on the `DsoRules` template having `dsoDelegate` as controller will introduce
+a new optional argument named `sv` to be used instead. This allows all SVs to exercise the choice by specifying their own party.
+Daml will reject a transaction where they try to specify any party other than their own with an authorization error.
+Once this CIP is implemented, SV applications will try to execute these choices by passing their own SV PartyId as controller.
```
```suggestion
```

seems to say nothing that wasn't said before.

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+e.g., if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.
+
+To minimize the impact, each SV application adds a random delay before each attempt and retries with a random backoff.
+A new metric named `splice_trigger_attempted_total` is added to monitor trigger contention and chose the right delay and backoff parameters.
```
```suggestion
A new metric named `splice_trigger_attempted_total` is added to monitor trigger failures caused by contention and help in choosing good delay and backoff parameters.
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+To minimize the impact, each SV application adds a random delay before each attempt and retries with a random backoff.
+A new metric named `splice_trigger_attempted_total` is added to monitor trigger contention and chose the right delay and backoff parameters.
+
+### Remove Delegate Election
```
```suggestion
### SV UI
```

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+
+## Specification
+
+### SV Application
```
I suggest splitting this into a Daml section that describes
1. the new controller argument
2. disabling delegate changes
and an SV app backend changes that describes that each SV attempts to submit with delay and backoff

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+A delegate performs critical network-related tasks, such as for example advancing mining rounds or executing accepted proposals.
+In the current version, there is no automatic fault tolerance to mitigate issues that may occur on the node. 
+SVs can vote on a new delegate, but this is a manual process, and the single point of failure is just moved to a different SV.
+A concerning problem observed multiple times is when a malfunctioning delegate causes mining rounds to stop progressing altogether. 
```
```suggestion
A concrete problem is rounds stop progressing when the delegate is down.
```
a bit more succinct and neutral

moritzkiefer-da commented on cip-0058/cip-0058.md:
```diff
@@ -0,0 +1,77 @@
+
+All SV applications attempting to complete tasks provides the right fault tolerance guarantees and introduces no correctness issues as commands are already all idempotent. 
+
+Random delays and backoffs are sufficient to minimize contention given that the delegate based choices do not require high throughput, 
```
```suggestion
Given that the current delegate-based choices don't require high throughput, e.g., round advancement happens only every 10 minutes, random delays and backoffs are sufficient to minimize contention.
```

moritzkiefer-da reviewed (approved):
Nice thanks! Let's address the remaining comments and then:
1. Create the splice branch with the code
2. Send the email
3. Make the PR

julientinguely-da committed:
final

hythloda committed:
cip 0058 Add IntellectEU as a Weight 1 Super Validator (#50)

* cip 0058 draft

* re-write

* add custody section

* add weights

* weights and provisional

* language in KMS section

* new template applied

* Update description for IntellectEU's Package Manager

* Update cip-0058.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0058.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update README.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Change "maintainer" to "contributor"

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* change "we" to "IntellectEU"

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Clarify that IEU must implement these mechanics for other SVs on the multi-tenant, not just for itself.

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Clarifying that the GSF will set up an extraBeneficiary in the name of the Super Validator

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Clarifying that the SV weight is assigned to each SV in a separate CIP

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Clarifying language

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Correcting the sequence of events

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Correcting the sequence of events

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Correcting the sequence of events

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Corrected the sequence of events

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Correcting the sequence of events

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Correcting sequence of events

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Additional correction to sequence

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
grammar fix

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
Update cip-0058.md (#56)

* Update cip-0058.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* adjust bullets

---------

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

hythloda committed:
licenses (#57)

julientinguely-da committed:
review wayne

Cmierly committed:
Add cip headers and seperate to folders (#54)

* Create cip-0003.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Rename cip-0003.md to cip-0002.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0002.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0003.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0006.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0007.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Delete cip-0006-0007 directory

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0009.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0008.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0006.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0010.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0011.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Delete cip-0009-0010-0011 directory

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0012.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0013.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0014.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0015.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0016.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0017.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Delete cip-0015-0016-0017 directory

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0018.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0019.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0020.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0021.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0022.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0024.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0025.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Delete cip-0024-0025 directory

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0032.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0033.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0034.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0036.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0038.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0039.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0040.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Add files via upload

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Delete cip-0032-0033-0034-0036-0038-0039-0040 directory

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0035.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Create cip-0042.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0006.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0007.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0006.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0008.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0009.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0010.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0011.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0012.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0013.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0013.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0014.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0015.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0016.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0017.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0018.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0019.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0020.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0021.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0022.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0024.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0025.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0032.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0033.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0034.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0035.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0035.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0036.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0038.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0039.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0040.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0042.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0001.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0002.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0003.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0006.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0007.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0008.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0009.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0010.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0011.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0012.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0013.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0014.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0015.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0016.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0017.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0018.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0019.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0020.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0021.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0022.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0024.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0025.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0032.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0033.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0034.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0035.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0036.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0037.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0038.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0039.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0040.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0041.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0042.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0043.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0044.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0045.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0046.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0048.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0049.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0051.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0054.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* Update cip-0057.md

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>

* links and readme

* update some dates

---------

Signed-off-by: Crystal Mierly <58894753+Cmierly@users.noreply.github.com>
Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Co-authored-by: Amanda L Martin <hythloda@gmail.com>

hythloda committed:
CIP 0055 Add Bitwave as a Weight 1 Super Validator (#44)

* cip-0055

* Clarified problems Bitwave is proposing to solve

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Reverted change in the last line

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Clarifying why Bitwave is a good match - outline text

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* improved clarity

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* fix typo

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* add prose to rational

* provisional language from new process

* provisional language from new process

* Fix formatting in cip-0055 documentation

* Update cip-0055.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* fix typo

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
cip-0059 draft (#52)

* cip-0059 draft

* Correct abstract to state "Woodside AI" and "weight 10" in place of "Monstera"

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Update cip-0059.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* approved

* changed relative (past 4 months) to absolute (early 2025) because this is a long-lived doc

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Removed redundant "working on" as it suggests they haven't started developing an investment thesis

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* clarity

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* clarified the separation between escrow process and minting process

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
fix 42 (#60)

hythloda committed:
Update cip-0042.md (#61)

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

hythloda committed:
Update cip-0057.md (#62)

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

hythloda committed:
CIP 0052 Add Monstera FZE as a Weight 3 Super Validator (#41)

* cip 0052 draft

* move to Specification section

* Remove "Proposed Changes" section from cip-0052.md

* Replace "commitment" with "rationale" to comply with CIP-0000

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* add underperform language

* Update GSF distribution and KPI conditions

* provisional language from new process

* new template

* Update cip-0052.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0052.md deliverable language

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0052.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0052.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0052.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Clarifying how the SV node joins when commitments are still pending

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
cip-0060 Add Zero Hash as SV of Weight 7.5 (#53)

* cip-0060 draft

* Update cip-0060.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* mechanincs

* Chris' feedback

* corrects table formatting

* Update cip-0060.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0060.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* approved cip

* Clarifying the process through which the SV node will join, if it joins before completing all milestones.

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

julientinguely-da committed:
add benchmark number

hythloda committed:
CIP 0053 - Add AngelHack as a Weight 2.5 SV (#42)

* cip 0053 draft

* move proposed changse

* Format into correct sections

* move sections

* add escrow language

* Feedback from Crhis

* provisional language from new process

* new template

* Update cip-0053.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0053.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Update cip-0053.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

* Corrected reference to "Canton" to refer to the "Splice" open source repositories

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarified that the SV applications are instances of the same app, run on each of the SV nodes

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
make seperate file for new cip. Restore 58

hythloda committed:
Merge branch 'main' into julien/delegateless-automation

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

hythloda committed:
Update cip-0058.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

matteolimberto-da committed:
add link to cip-0056 in README (#64)

* add link to cip-0056 in README

* Update README.md

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

---------

Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Co-authored-by: Amanda L Martin <hythloda@gmail.com>

waynecollier-da committed:
Correct "Governance" to "Standards Track"

This CIP describes code changes to the standard implementation of Splice. It does not introduce changes to the governance process; it only changes how the implementation of a decision is submitted (end submission by a fixed delegate, implement random submission).

Co-authored-by: Stas German (SBI) <stasge+sbi@sbisecsol.com>
Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
Update cip-0055.md (#68)

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

julientinguely-da committed:
add missing parts from Stas review

julientinguely-da committed:
nits

julientinguely-da committed:
Update cip-00XX/cip-00XX.md

Co-authored-by: Stas German (SBI) <stasge+sbi@sbisecsol.com>
Signed-off-by: Julien Tinguely <128390688+julientinguely-da@users.noreply.github.com>

hythloda committed:
assign number

hrischuk-da committed:
CIP 0062-Synchronizer-Migration-to-Canton-3-3.md (#66)

* Create cip-XXXX-Synchronizer-Migration-to-Canton-3-3.md

* Added changelog entry

* Update cip-XXXX-Synchronizer-Migration-to-Canton-3-3/cip-XXXX-Synchronizer-Migration-to-Canton-3-3.md

Co-authored-by: moritzkiefer-da <45630097+moritzkiefer-da@users.noreply.github.com>
Signed-off-by: hrischuk-da <curtis.hrischuk@digitalasset.com>

* Updates based on feedback - May 20

* assign number

* Approved CIP

---------

Signed-off-by: hrischuk-da <curtis.hrischuk@digitalasset.com>
Co-authored-by: moritzkiefer-da <45630097+moritzkiefer-da@users.noreply.github.com>
Co-authored-by: Amanda L Martin <hythloda@gmail.com>

waynecollier-da committed:
noted the importance of this CIP to governance.

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
Clarified that submissions are only after 2/3 threshold is reached.

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
Clarifications to the language

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

moritzkiefer-da commented on cip-0064/cip-0064.md:
```diff
@@ -15,7 +15,7 @@
 
 This CIP proposes to replace the elected SV delegate mechanism, to improve fault tolerance and more fully automate decentralized operations.  
-Instead of relying on a single delegate, we propose an approach where the backend automation for each instance of the Super Validator application, (one instance operating on each of the different Super Validator nodes), attempts to exercise delegate choices. 
+Instead of relying on a single delegate, we propose an approach where the backend automation for each instance of the Super Validator application, (one instance operating on each of the different Super Validator nodes), attempts to submit actions that have reached the 2/3 voting threshold in favor. 
```
@waynecollier-da that is change is partially incorrect. E.g., round advancement does not have an explicit 2/3 voting threshold because the condition is just that the time has advanced fast enough which is purely on ledger.

It really is just all delegate choices. Some of those delegate choices just require the delegate to pass in vote contracts from at least 2/3 of SVs if that's the only way it can be executed safely.

waynecollier-da replied:
ok, I need to understand this better, then. I'll add it to one of our discussions.

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

waynecollier-da committed:
clarification

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
Merge branch 'main' into julien/delegateless-automation

Signed-off-by: Amanda L Martin <hythloda@gmail.com>

waynecollier-da committed:
Wording

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
cip-0063 Add Kaiko as a Super Validator of Weight 6.5 (#69)

* draft of cip-0063

* Approved

* small grammatical fix

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

* Correct spelling error

Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>

---------

Signed-off-by: Amanda L Martin <hythloda@gmail.com>
Signed-off-by: Wayne Collier <wayne.collier@digitalasset.com>
Co-authored-by: Wayne Collier <wayne.collier@digitalasset.com>

hythloda committed:
Approved

hythloda committed:
Merge branch 'main' into julien/delegateless-automation

Signed-off-by: Amanda L Martin <hythloda@gmail.com>
